feat(buzzer): Add initial support for the buzzer driver

AnthonyGrondin opened the pull request:
Provides a driver for a piezo-electric buzzer by abstracting LEDC and offering a user-friendly API for a buzzer.

It's currently a lot of work and not user friendly to manage both the channel and the timer to drive a piezo-electric buzzer. This PR solves this issues by providing a buzzer module that is more user-friendly to use

## Thank you!

Thank you for your contribution.
Please make sure that your submission includes the following:

### Must

- [x] The code compiles without `errors` or `warnings`.
- [x] All examples work.
- [x] `cargo fmt` was run.
- [x] Your changes were added to the `CHANGELOG.md` in the proper section.
- [x] You updated existing examples or added examples (if applicable).
- [x] Added examples are checked in CI

### Nice to have

- [x] You add a description of your work to this PR.
- [x] You added proper docs for your newly added features and code.

### TODO
- [x] Add volume support through duty
- [x] Add missing examples

bjoernQ commented:
Looks nice and useful but maybe it's not really something to be part of the core-HAL. Instead, I would say it's similar to the SmartLED implementation.

Maybe we could have this and the SmartLED implementation in a crate like `esp-hal-extras` (name is debatable - I'm really bad at coming up with names)

For the future there are probably even more functionalities which can go into `esp-hal-extras`

Just my initial thoughts - happy to hear different opinions

AnthonyGrondin commented:
I've moved buzzer into its own crate in `esp-hal-buzzer` and moved `esp-hal-smartled` into `esp-hal-extras/`.

I'm currently have a GPIO generic error when not using a volume pin due to the type constraints. I'm not sure how to fix it.

I didn't add examples for the other SoCs. I did a quick testing on `esp32c3` and the frequency wasn't correct.

jessebraham commented:
Sorry to be a pain, but I'd prefer if the directory was just called `extras` I think, please. When we start adding HIL testing and stuff we will likely need some simple libraries and/or binary crates to support that, and who knows what else we will come up with in the future, so I don't want the directory's name to imply these packages are tightly coupled to `esp-hal-common` (even though this is currently the case).

Haven't had a chance to look at the actual library yet but will hopefully get to that soon.

AnthonyGrondin commented:
@jessebraham This is still in Draft process, but the CI seems to be failing on the MSRV check, and I'm not sure why.

AnthonyGrondin commented:
TODO:

- [ ] Add example for `esp32s2`
- [ ] Add example for `esp32c6`
- [ ] Add example for `esp32h2`
- [ ] Fix issue for RISC-V targets

I've noticed a bug on RISVC targets, on both `esp32c2` and `esp32c3`, where the frequency played is way too high.

I've tested on both `esp32s3` and `esp32` and it seems to work fine on my side.

jessebraham commented:
Not sure what was going on with CI, but I've re-run the failing job and it seems to be happy now, so all good there.

bjoernQ commented:
The problem on H2 _might_ be related to https://github.com/esp-rs/esp-hal/issues/766

AnthonyGrondin commented:
Ope. I forgit I was using my main for an important PR oopsie.
